CPAC/pipeline/nodeblock.py

```diff
@@ -0,0 +1,117 @@
+"""Class and decorator for NodeBlock functions"""
+from typing import Callable, List, Union, Optional, Dict, Any
+
+
+class NodeBlockFunction(object):
+    """
+    Stores a reference to the nodeblock function and all of its meta-data.
+    """
+
+    def __init__(
+            self,
+            func: Callable,
+            name: Optional[str] = None,
+            config: Optional[List[str]] = None,
+            switch: Optional[Union[List[str], List[List[str]]]] = None,
+            option_key: Optional[Union[str, List[str]]] = None,
+            option_val: Optional[Union[str, List[str]]] = None,
+            inputs: Optional[List[Union[str, list, tuple]]] = None,
+            outputs: Optional[Union[List[str], Dict[str, Any]]] = None
```

Comment:
It took me awhile to find this (it was [in a comment thread in an OHBM abstract](https://docs.google.com/document/d/1fMwWUiaoDcuE271fDl0KsqK8Ne3it3OqKewkV4kk17Y/edit?disco=AAAAmGuwDoY)):


> A value can be the string “None” instead of one of the specified types, but each must be specified.

― abstract

> > Now we have to specify "None". Would it be too dangerous / to implicit to allow excluding entities that are "None"? (I think maybe)

― @shnizzedy 

> > I think it would be too dangerous, yes.

― @gkiar

Obviously we'll update the poster with the decorators, but I think we need to make all of these required.

I do think we could make them nullable though, so we could specify whatever `=None` instead of `="None"` unless anyone has a good argument against that (@sgiavasis? @gkiar?)

Reply:
I don't fully get the reasoning here, if a nodeblock has no switch why does the user explicitly have to type switch=None ? How is this more secure?

Reply:
Same goes for `name`, right now if you omit it it will get the function name which almost always makes sense

Reply:
Re: name, I'd vote for making it optional and defaulting to something like f`{func.__name__}_nodeblock` instead of just `func.__name__` so working paths and crashfiles are immediately apparent whether it's the function or the nodeblock.

Re: the rest of them being optional, I can see benefits to forcing developers to be explicit, so, for example, they don't forget to plug in a switch, but I can also appreciate avoiding a bunch of explicit `=Nones` throughout the code base that could have just been defaults. I don't really care either way; whatever we do here (@sgiavasis @gkiar), we'll just document that in the poster (and the dev docs)

Reply:
An addition to that would be adding `_nodeblock` every time (also to the name strings) for consistency.

Reply:
…iff the given name doesn't already end with (contain?) "`nodeblock`" 

Reply:
I don't remember deliberately choosing it to be explicit- I'm fine with making it optional.

Reply:
I believe it was as an insulation against typos and forgetting. I'm fine with using default (nullable) values now that we're using the decorator method for nodeblock definition.